src/software/amazon/ion/impl/IonReaderBinaryRawX.java

```diff
@@ -888,175 +888,210 @@ protected final BigInteger readBigInteger(int len, boolean is_negative) throws I
             // for the second byte we shift our eariler bits just as much,
             // but there are fewer of them there to shift
+
             if ((b = read()) < 0) throwUnexpectedEOFException();
             retvalue = (retvalue << 7) | (b & 0x7F);
             if ((b & 0x80) != 0) break done;
             // for the rest, they're all the same
+
             if ((b = read()) < 0) throwUnexpectedEOFException();
             retvalue = (retvalue << 7) | (b & 0x7F);
             if ((b & 0x80) != 0) break done;
             // for the rest, they're all the same
+
             if ((b = read()) < 0) throwUnexpectedEOFException();
             retvalue = (retvalue << 7) | (b & 0x7F);
             if ((b & 0x80) != 0) break done;
             // for the rest, they're all the same
+
             if ((b = read()) < 0) throwUnexpectedEOFException();
             retvalue = (retvalue << 7) | (b & 0x7F);
             if ((b & 0x80) != 0) break done;
-            // if we get here we have more bits than we have room for :(
-            throwIntOverflowExeption();
+
+            // Don't support anything above a 5 bytes VarUInt for now, see https://github.com/amzn/ion-java/issues/146
+            throwVarIntOverflowException();
         }
         if (is_negative) {
             retvalue = -retvalue;
         }
-        return retvalue;
-    }
-    protected final long readVarLong() throws IOException
-    {
-        long    retvalue = 0;
-        boolean is_negative = false;
-        int     b;
-        // synthetic label "done" (yuck)
-done:   for (;;) {
-            // read the first byte - it has the sign bit
-            if ((b = read()) < 0) throwUnexpectedEOFException();
-            if ((b & 0x40) != 0) {
-                is_negative = true;
-            }
-            retvalue = (b & 0x3F);
-            if ((b & 0x80) != 0) break done;
-            // for the second byte we shift our eariler bits just as much,
-            // but there are fewer of them there to shift
-            if ((b = read()) < 0) throwUnexpectedEOFException();
-            retvalue = (retvalue << 7) | (b & 0x7F);
-            if ((b & 0x80) != 0) break done;
-            // for the rest, they're all the same
-            for (;;) {
-                if ((b = read()) < 0) throwUnexpectedEOFException();
-                if ((retvalue & 0xFE00000000000000L) != 0) throwIntOverflowExeption();
-                retvalue = (retvalue << 7) | (b & 0x7F);
-                if ((b & 0x80) != 0) break done;
-            }
-        }
-        if (is_negative) {
-            retvalue = -retvalue;
+
+        int retValueAsInt = (int) retvalue;
+        if(retvalue != ((long)retValueAsInt)) {
+            throwVarIntOverflowException();
         }
-        return retvalue;
+
+        return retValueAsInt;
     }
+
     /**
      * Reads an integer value, returning null to mean -0.
      * @throws IOException
      */
     protected final Integer readVarInteger() throws IOException
```

Comment:
should we consider having a single impl for both readVarInt() and readVarInteger()?